blogpost: Multi-version upgrades (#5723)
sourcegraph/about: content/blogposts/2022/multi-version-upgrades.md

---
title: "Multi-version upgrades"

> description: "Placeholder."

Reviewer: Bringing attention to this to make sure we don't forget updating this as well as the `heroImage` and `socialImage` before we merge this.

Author: Updated both to https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/mvu-hero.png.

authors:
  - name: Eric Fritz
    url: https://eric-fritz.com

> publishDate: 2022-09-09T18:00+02:00

Reviewer 2: Just adding a note to update publish date when we're ready

tags: [blog]
slug: multi-version-upgrades
heroImage: placeholder
socialImage: placeholder
published: true
---

It's finally happened.

The most obnoxious restriction (by a very wide margin) related to administration of a Sourcegraph instance no longer applies. Sourcegraph instances (v3.20 or newer) are now able to upgrade directly to a future release, without the requirement to hit every minor release along the way.

A recent peek at our customer instance distribution showed that 70% of our self-hosted customers are on a "historic" release of Sourcegraph (older than two months). Larger customer instances also tend to approach upgrades a bit more cautiously, adding significant weight to the use of older releases. Note that [Sourcegraph v3.20](https://about.sourcegraph.com/blog/release/3.20) was released fairly _early_ in the most recent pandemic era, and we've spent the last two years of that era shipping improvements and building new features and workflows for an on-prem audience that was largely unable to benefit from the effort.

> [Multi-version upgrades](https://docs.sourcegraph.com/admin/updates#multi-version-upgrades), contrasting to the "standard upgrade" approach, use the [`migrator` tool](introducing-migrator-service) to perform the sequence of schema and in-place data migrations that _would happen_ if the user were to perform a chain of upgrades that hit every minor release on the way to their target. This tool makes upgrading directly to our [4.0 release](TODO:FILLMEIN) a technical feasibility for the vast majority of our customers.

Author: Ensure we update the placeholder link targett.

Reviewer 2: https://about.sourcegraph.com/blog/release/4.0

![treeview](temp/treeview.png)

This process didn't cause any clearly attributable problems for years. Around the time we started a serious hunt-to-kill effort for flaky tests in our CI pipelines, we started also [seeing initialization timeouts](https://github.com/sourcegraph/sourcegraph/issues/5721#issuecomment-540855232) we could attribute to reading and applying every single migration we've ever defined back-to-back.

Because this problem would only get worse with time, we began to squash migration definitions by replacing the set of definitions at the head of the sequence with a single (yet equivalent) migration file. For Sourcegraph v3.9.0, we [squashed definitions](https://github.com/sourcegraph/sourcegraph/pull/5967) that were also defined in Sourcegraph v3.7.0 into to a single file. This cut startup time with an empty database from 20s to 5s, and also reduced the chance of hitting the startup timeout in our testing environment.

![simple squash](temp/simple-squash.png)

> Note that we are unable to squash migration definitions that are not also defined in the source version of an upgrade. Doing so would make it impossible for the previous release to determine _which_ migration definitions to apply. Squashing in the manner described above guarantees that none of the migration definitions in the upgrade path from the previous minor version would be altered.

Author: The sentence is correct but really clunky. Does `Note that we are unable to squash migration definitions that are undefined in the source version of an upgrade.` read better?

> Note that we are unable to squash migration definitions that are not also defined in the source version of an upgrade. Doing so would make it impossible for the previous release to determine _which_ migration definitions to apply. Squashing in the manner described above guarantees that none of the migration definitions in the upgrade path from the previous minor version would be altered.

Reviewer: Oh, I see now what you meant there. Then the original phrasing is more precise, but I find your suggested alternative easier to understand.

> Note that we are unable to squash migration definitions that are not also defined in the source version of an upgrade. Doing so would make it impossible for the previous release to determine _which_ migration definitions to apply. Squashing in the manner described above guarantees that none of the migration definitions in the upgrade path from the previous minor version would be altered.

Author: ```suggestion
Note that we are unable to squash migration definitions that are undefined in the source version of an upgrade. Doing so would make it impossible for the previous release to determine _which_ migration definitions to apply. Squashing in the manner described above guarantees that none of the migration definitions in the upgrade path from the previous minor version would be altered.
```

We threw the **Wrench of Privileged Migrations** into the pile when we started marking certain SQL commands as requiring elevated permissions. Migration definitions that contain such a SQL command and don't mark it (or mark it but do not contain such a SQL command) fail static validation. Migration definitions squashed away before the introduction of this change fail these validation checks when added to the unified migration graph. In order to bring these definitions in line with our current assumptions, we spot-rewrite the metadata of such migration definitions as we stitch the graph together. We do this for several `frontend` ([1528395717](https://github.com/sourcegraph/sourcegraph/pull/11457), [1528395764](https://github.com/sourcegraph/sourcegraph/pull/15641), and [1528395953](https://github.com/sourcegraph/sourcegraph/pull/28387)), `codeintel` ([1000000003](https://github.com/sourcegraph/sourcegraph/pull/15641) and [1000000020](https://github.com/sourcegraph/sourcegraph/pull/25199)), and `codeinsights` ([1000000001](https://github.com/sourcegraph/sourcegraph/pull/18083) and [1000000027](https://github.com/sourcegraph/sourcegraph/pull/30781)) migration definitions.

> We deal with the **Wrench of Concurrent Indexes** in a similar way. Migration definitions squashed away before the introduction of our special-casing of concurrent index creation will miss the markers in their metadata. In this case we only have to spot rewrite migrations for the `frontend` schema, but there were a significant nubmer of them ([1528395696](https://github.com/sourcegraph/sourcegraph/pull/12591), [1528395707](https://github.com/sourcegraph/sourcegraph/pull/13217), [1528395708](https://github.com/sourcegraph/sourcegraph/pull/13217), [1528395736](https://github.com/sourcegraph/sourcegraph/pull/14848), [1528395797](https://github.com/sourcegraph/sourcegraph/pull/19053), [1528395877](https://github.com/sourcegraph/sourcegraph/pull/24798), [1528395878](https://github.com/sourcegraph/sourcegraph/pull/24798), [1528395886](https://github.com/sourcegraph/sourcegraph/pull/24969), [1528395887](https://github.com/sourcegraph/sourcegraph/pull/24969), [1528395888](https://github.com/sourcegraph/sourcegraph/pull/24969), [1528395893](https://github.com/sourcegraph/sourcegraph/pull/25366), [1528395894](https://github.com/sourcegraph/sourcegraph/pull/25360), [1528395896](https://github.com/sourcegraph/sourcegraph/pull/25405), [1528395897](https://github.com/sourcegraph/sourcegraph/pull/25490), [1528395899](https://github.com/sourcegraph/sourcegraph/pull/25520), [1528395900](https://github.com/sourcegraph/sourcegraph/pull/25520), [1528395935](https://github.com/sourcegraph/sourcegraph/pull/27359), [1528395936](https://github.com/sourcegraph/sourcegraph/pull/27359), and [1528395954](https://github.com/sourcegraph/sourcegraph/pull/28387)).

Reviewer 2: ```suggestion
We deal with the **Wrench of Concurrent Indexes** in a similar way. Migration definitions squashed away before the introduction of our special-casing of concurrent index creation will miss the markers in their metadata. In this case we only have to spot rewrite migrations for the `frontend` schema, but there were a significant number of them ([1528395696](https://github.com/sourcegraph/sourcegraph/pull/12591), [1528395707](https://github.com/sourcegraph/sourcegraph/pull/13217), [1528395708](https://github.com/sourcegraph/sourcegraph/pull/13217), [1528395736](https://github.com/sourcegraph/sourcegraph/pull/14848), [1528395797](https://github.com/sourcegraph/sourcegraph/pull/19053), [1528395877](https://github.com/sourcegraph/sourcegraph/pull/24798), [1528395878](https://github.com/sourcegraph/sourcegraph/pull/24798), [1528395886](https://github.com/sourcegraph/sourcegraph/pull/24969), [1528395887](https://github.com/sourcegraph/sourcegraph/pull/24969), [1528395888](https://github.com/sourcegraph/sourcegraph/pull/24969), [1528395893](https://github.com/sourcegraph/sourcegraph/pull/25366), [1528395894](https://github.com/sourcegraph/sourcegraph/pull/25360), [1528395896](https://github.com/sourcegraph/sourcegraph/pull/25405), [1528395897](https://github.com/sourcegraph/sourcegraph/pull/25490), [1528395899](https://github.com/sourcegraph/sourcegraph/pull/25520), [1528395900](https://github.com/sourcegraph/sourcegraph/pull/25520), [1528395935](https://github.com/sourcegraph/sourcegraph/pull/27359), [1528395936](https://github.com/sourcegraph/sourcegraph/pull/27359), and [1528395954](https://github.com/sourcegraph/sourcegraph/pull/28387)).
```

Reviewer 2: nubmer -> number

---
title: 'Multi-version upgrades'
description: 'Sourcegraph administration just got an order-of-magnitude quality of life improvement. Using a Sourcegraph 4.0 (or later) migrator utility, you can upgrade your on-prem installation of Sourcegraph from 3.20 directly to Sourcegraph 4.0 and beyond.'
authors:
  - name: Eric Fritz
    url: https://eric-fritz.com

> publishDate: 2022-10-19T10:00-07:00

Author: ```suggestion
publishDate: 2022-10-20T10:00-07:00
```

@iskyOS Will leave this to you to choose a publish date :)

Each migration exists on-disk as a pair of SQL files: one for the upgrade direction, and one for the downgrade direction. Each migration is defined in the same directory and ordered absolutely by a sequential numeric prefix. Unit tests ensure that the sequence has no gaps and that no migrations define the same prefix. The order of migration application is thereby determined by lexicographical ordering of the migration filenames.

<Figure src="https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/treeview.png" alt="List of the (old) migrations directory." />

This process didn't cause any clearly attributable problems for years. Around the time we started a serious hunt-to-kill effort for flaky tests in our CI pipelines, we started also [seeing initialization timeouts](https://github.com/sourcegraph/sourcegraph/issues/5721#issuecomment-540855232) we could attribute to reading and applying every single migration we've ever defined back-to-back.

> Because this problem would only get worse with time, we began to squash migration definitions by replacing the set of definitions at the head of the sequence with a single (yet equivalent) migration file. For Sourcegraph v3.9.0, we [squashed definitions](https://github.com/sourcegraph/sourcegraph/pull/5967) that were also defined in Sourcegraph v3.7.0 into to a single file. This cut startup time with an empty database from 20s to 5s, and also reduced the chance of hitting the startup timeout in our testing environment.

Reviewer 2: ```suggestion
Because this problem would only get worse with time, we began to squash migration definitions by replacing the set of definitions at the head of the sequence with a single (yet equivalent) migration file. For Sourcegraph v3.9.0, we [squashed definitions](https://github.com/sourcegraph/sourcegraph/pull/5967) that were also defined in Sourcegraph v3.7.0 into a single file. This cut startup time with an empty database from 20s to 5s and also reduced the chance of hitting the startup timeout in our testing environment.
```

<Figure src="https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/tree.png" alt="File tree view of the (updated) migrations directory." />

In what is likely the largest departure from golang-migrate, we allow migrations to [define multiple parents](https://github.com/sourcegraph/sourcegraph/pull/30664). This allows our migration definitions to form a [directed acyclic graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph) (instead of an absolutely ordered list). Note that in the image above our migrations no longer form a gapless sequence, and a [topological sort](https://en.wikipedia.org/wiki/Topological_sorting) over the migration graph is done to determine an application order.

> Well that seemed like a bunch of work for no payoff. **Why do that at all?** Short answer: engineering ergonomics. When we had a gapless sequence we would **frequently** run into merge conflicts when two engineers add migrations around the same time. Since each branch did not contain the other's migration, they would both choose the same sequence number. Because of limitations in our CI (no build queues + no non-stale base branch merge requirement), this conflict would, the majority of the time, be caught only after the both branchs have been merged and validation errors turned the `main` branch to go red.

Reviewer 2: ```suggestion
Well that seemed like a bunch of work for no payoff. **Why do that at all?** Short answer: engineering ergonomics. When we had a gapless sequence we would **frequently** run into merge conflicts when two engineers add migrations around the same time. Since each branch did not contain the other's migration, they would both choose the same sequence number. Because of limitations in our CI (no build queues + no non-stale base branch merge requirement), this conflict would, the majority of the time, be caught only after both branches have been merged and validation errors turned the `main` branch to go red.
```

![graph structures](https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/mvu-graph-structures.png)

## The pile of wrenches

The evolution of our migration definitions over time has created a number of issues to overcome for multi-version upgrades. The biggest issue comes from periodically squashing migration definitions, which erases information necessary to the upgrade process. In the image above, there's no valid way to upgrade from 3.36 -> 3.39 in one step. The squashed migration contains _both_ the SQL that's already been ran by the 3.36 installation, as well as SQL defined in 3.37 that has yet to be applied. As we cannot partially apply a single migration, we cannot support this upgrade path.

To get around this, we need to undo the squash operation over time to get a single, consistent "stitched" migration graph.

> ![timeline](temp/timeline.png)

Reviewer 3: This isn't loading, needs to be uploaded to gcloud @efritz

Reviewer 2: We also have new guidance on embedding images using <Figure>: https://about.sourcegraph.com/blog/starter-pack#video

Reviewer 2: ```suggestion
<Figure
    src="blog/image name"
    alt=""
    caption=""
  />
```

Author: Updated these.

Author: @jdorfman I think you were looking at an old version.

> <Figure src="temp/timeline.png" alt="timeline" />

Reviewer 2: @efritz What image should this be?

Author: https://github.com/sourcegraph/about/blob/8eda68889b3f6db9cb4638003ee6ca64e4a7c7b5/public/blog/temp/timeline.png

Author: ```suggestion
<Figure src="https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/timeline.png" alt="timeline" />
```

To solve this, we also [run the out-of-band migrations in the migrator](https://github.com/sourcegraph/sourcegraph/issues/38094), interleaved with the schema migrations.

<Figure src="https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/mvu-oobmigrations.png" alt="How out-of-band migrations complicate multi-version upgrades." />

In the example above, upgrading from v3.37 to v3.40 would require the completion of out-of-band migrations #1-#3. To make things a bit more difficult, the out-of-band migration code is only guaranteed to run against the database schema that existed within its lifetime. We cannot invoke out-of-band migration #3 with the v3.37.0 database schema, for example. This becomes a scheduling problem in which we upgrade the schema to a certain version, run out-of-band migrations to completion at that point, and continue the process.

## Conclusion

> We've continued to iterate on our schema and data migration infrastructure to remove entire classes of pain from our customers. We know that Sourcegraph administration is not a cakewalk, and we are taking significant strides to decrease maintenance burden. Keep your eyes peeled for future improvements in this area.

Reviewer 2: ```suggestion
We've continued to iterate on our schema and data migration infrastructure to remove entire classes of pain from our customers, and we are taking significant strides to decrease maintenance burden. Keep your eyes peeled for future improvements in this area.
```
